runtime/drive_handler.go

```diff
@@ -96,32 +96,79 @@ type StubDriveHandler struct {
 // Reserve pops a unused stub drive and returns a MountableStubDrive that can be
 // mounted with the provided options as the patched drive information.
 func (h *StubDriveHandler) Reserve(
+	requestCtx context.Context,
 	id string,
 	hostPath string,
 	vmPath string,
 	filesystemType string,
 	options []string,
-) (MountableStubDrive, error) {
+	driveMounter drivemount.DriveMounterService,
+	machine firecracker.MachineIface,
+) error {
 	h.mu.Lock()
 	defer h.mu.Unlock()
 	if len(h.freeDrives) == 0 {
-		return nil, ErrDrivesExhausted
+		return ErrDrivesExhausted
+	}
+	if _, ok := h.usedDrives[id]; ok {
+		// This case means that driver wasn't released or removed properly
+		return fmt.Errorf("drive with ID %s already in use, a previous attempt to remove it may have failed", id)
 	}
 
 	freeDrive := h.freeDrives[0]
 	options, err := setReadWriteOptions(options, freeDrive.driveMount.IsWritable)
 	if err != nil {
-		return nil, err
+		return err
 	}
 
-	h.freeDrives = h.freeDrives[1:]
-	h.usedDrives[id] = freeDrive
-	return freeDrive.withMountConfig(
+	stubDrive := freeDrive.withMountConfig(
```

Comment:
I think we should probably not use a separate `stubDrive` variable here and instead do `freeDrive = freeDrive.withMountConfig(...)`. This will ensure we get all the relevent state stored in the map and should also fix some of the unit test failures you're getting right now (I suspect anyways).